Add core coordination algorithm for cluster state publishing

ywelsch opened the pull request:
Adds the core coordination algorithm that guarantees safety for cluster state publishing. This is a straight-forward port of the transition rules from the [formal model](url
https://github.com/elastic/elasticsearch-formal-models/blob/master/ZenWithTerms/tla/ZenWithTerms.tla) and goes hand-in-hand with the PR https://github.com/elastic/elasticsearch-formal-models/pull/37. 

Relates to #32006

DaveCTurner commented on server/src/main/java/org/elasticsearch/cluster/coordination/ApplyCommit.java:
```diff
@@ -0,0 +1,49 @@
+
+import java.io.IOException;
+
+public class ApplyCommit extends TermVersionRequest {
```
Classes derived from `TransportRequest` seem to be named with a `...Request` suffix. `ApplyCommitRequest` here?

ywelsch replied:
sure, fixed in 34567f669f7

DaveCTurner commented on server/src/test/java/org/elasticsearch/cluster/coordination/CoordinationStateTests.java:
```diff
@@ -0,0 +1,670 @@
+
+public class CoordinationStateTests extends ESTestCase {
+
+    DiscoveryNode node1;
```
These fields can all be `private`. Additionally, `initialStateNode{2,3}` and `cs3` are only used in `setupNodes()`.

ywelsch replied:
fixed in 6611e58a67d (cs3 is now used)

DaveCTurner commented on server/src/test/java/org/elasticsearch/cluster/coordination/CoordinationStateTests.java:
```diff
@@ -0,0 +1,670 @@
+import static org.hamcrest.Matchers.containsString;
+import static org.hamcrest.Matchers.equalTo;
+
+public class CoordinationStateTests extends ESTestCase {
```
I think we should have a test that runs the state machines in parallel, passing messages back and forth, and asserting that the committed states remain consistent. This might want to wait until the PR on which I'm working which introduces a deterministic task queue, so maybe just a `//TODO` for now.

DaveCTurner commented on server/src/main/java/org/elasticsearch/cluster/coordination/CoordinationState.java:
```diff
@@ -0,0 +1,485 @@
+     * @throws CoordinationStateRejectedException if the arguments were incompatible with the current state of this object.
+     */
+    public Join handleStartJoin(StartJoinRequest startJoinRequest) {
+        if (startJoinRequest.getTerm() <= getCurrentTerm()) {
```
Tests all pass if this is `<` instead of `<=`.

ywelsch replied:
Did you run the tests multiple times? Both testStartJoinBeforeBootstrap and testStartJoinAfterBootstrap fail approx. 30% of the times if this is changed from `<=` to `<`.

DaveCTurner replied:
Only a couple of times, not enough clearly.

ywelsch replied:
I've pushed 3bec606f179 🗡

DaveCTurner commented on server/src/main/java/org/elasticsearch/cluster/coordination/CoordinationState.java:
```diff
@@ -0,0 +1,485 @@
+        logger.debug("handleJoin: added join {} from [{}] for election, electionWon={} lastAcceptedTerm={} lastAcceptedVersion={}", join,
+            join.getSourceNode(), electionWon, lastAcceptedTerm, getLastAcceptedVersion());
+
+        if (electionWon && prevElectionWon == false) {
```
Tests all pass with `if (electionWon) {`

ywelsch replied:
fixed in 51192562026 (which also found another small copy-paste error)

DaveCTurner commented on server/src/main/java/org/elasticsearch/cluster/coordination/CoordinationState.java:
```diff
@@ -0,0 +1,485 @@
+                " of join higher than current last accepted term " + lastAcceptedTerm);
+        }
+
+        if (join.getLastAcceptedTerm() == lastAcceptedTerm && join.getLastAcceptedVersion() > getLastAcceptedVersion()) {
```
Tests all pass with `if (join.getLastAcceptedVersion() > getLastAcceptedVersion()) {`

ywelsch replied:
fixed in 8d2fc8cddd6 and acad072945f

DaveCTurner commented on server/src/main/java/org/elasticsearch/cluster/coordination/CoordinationState.java:
```diff
@@ -0,0 +1,485 @@
+                "incoming term " + join.getTerm() + " does not match current term " + getCurrentTerm());
+        }
+
+        if (startedJoinSinceLastReboot == false) {
```
Tests pass with `if (false) {`

ywelsch replied:
fixed in 8b7af72f657

DaveCTurner commented on server/src/main/java/org/elasticsearch/cluster/coordination/CoordinationState.java:
```diff
@@ -0,0 +1,485 @@
+        }
+
+        if (clusterState.getLastAcceptedConfiguration().equals(getLastAcceptedConfiguration()) == false
+            && getLastCommittedConfiguration().equals(getLastAcceptedConfiguration()) == false) {
```
Tests pass with `if (getLastCommittedConfiguration().equals(getLastAcceptedConfiguration()) == false) {`

ywelsch replied:
fixed in 708ae8fd098

DaveCTurner commented on server/src/main/java/org/elasticsearch/cluster/coordination/CoordinationState.java:
```diff
@@ -0,0 +1,485 @@
+            throw new CoordinationStateRejectedException("incoming term " + clusterState.term() + " does not match current term " +
+                getCurrentTerm());
+        }
+        if (clusterState.term() == getLastAcceptedTerm() && clusterState.version() <= getLastAcceptedVersion()) {
```
Tests pass with:
- `if (clusterState.version() <= getLastAcceptedVersion()) {`
- `if (clusterState.version() < getLastAcceptedVersion()) {`
- `if (clusterState.term() == getLastAcceptedTerm()) {`

ywelsch replied:
fixed in 0d84428d208

DaveCTurner commented on server/src/main/java/org/elasticsearch/cluster/coordination/CoordinationState.java:
```diff
@@ -0,0 +1,485 @@
+    }
+
+    public boolean isElectionQuorum(VoteCollection votes) {
+        return votes.isQuorum(getLastCommittedConfiguration()) && votes.isQuorum(getLastAcceptedConfiguration());
```
Tests pass with
- `return votes.isQuorum(getLastCommittedConfiguration());`
- `return votes.isQuorum(getLastAcceptedConfiguration());`

ywelsch replied:
fixed in ab02587ec40

DaveCTurner commented on server/src/main/java/org/elasticsearch/cluster/coordination/CoordinationState.java:
```diff
@@ -0,0 +1,485 @@
+    }
+
+    public boolean isPublishQuorum(VoteCollection votes) {
+        return votes.isQuorum(getLastCommittedConfiguration()) && votes.isQuorum(lastPublishedConfiguration);
```
Tests pass with
- `return votes.isQuorum(getLastCommittedConfiguration());`
- `return votes.isQuorum(lastPublishedConfiguration);`

ywelsch replied:
fixed in 8ee68b9d79a

ywelsch commented:
@DaveCTurner Thanks for the great review. I've run a code coverage tool on this which found one more missing test case, fixed in 9b2ad945ae2. I have addressed all but one comment (the randomized testing), but I think this is ready for another review. I will check in the meanwhile how tricky it is to add these tests.

DaveCTurner commented on server/src/main/java/org/elasticsearch/cluster/coordination/Join.java:
```diff
@@ -0,0 +1,127 @@
+/**
+ * Triggered by a {@link StartJoinRequest}, instances of this class represent join votes,
+ * and have a source and target node. The source node is the node that provides the vote,
+ * and the target node is the node for which this vote is casted. A node will only cast
```
Nit: `casted` should read `cast`.
